arangod/Aql/Functions.cpp

```diff
@@ -8848,6 +8856,148 @@ AqlValue Functions::MakeDistributeGraphInput(arangodb::aql::ExpressionContext* e
+                       AstNode const& node,
+                       VPackFunctionParameters const& parameters,
+                       F&& decayFuncFactory) {
+
+  AqlValue const& argValue = extractFunctionParameterValue(parameters, 0);
+  AqlValue const& originValue = extractFunctionParameterValue(parameters, 1);
+  AqlValue const& scaleValue = extractFunctionParameterValue(parameters, 2);
+  AqlValue const& offsetValue = extractFunctionParameterValue(parameters, 3);
+  AqlValue const& decayValue = extractFunctionParameterValue(parameters, 4);
+
+  // check type of arguments
+  if ((!argValue.isRange() && !argValue.isArray() && !argValue.isNumber()) ||
+    !originValue.isNumber() || !scaleValue.isNumber() ||
+    !offsetValue.isNumber() || !decayValue.isNumber()) {
+
+    registerInvalidArgumentWarning(expressionContext,
+                                   getFunctionName(node).c_str());
+    return AqlValue(AqlValueHintNull());
+  }
+
+  // extracting values
+  bool failed;
+  bool error = false;
+  double origin = originValue.toDouble(failed);
+  error |= failed;
+  double scale = scaleValue.toDouble(failed);
+  error |= failed;
+  double offset = offsetValue.toDouble(failed);
+  error |= failed;
+  double decay = decayValue.toDouble(failed);
+  error |= failed;
+
+  if (error) {
+    registerWarning(expressionContext,
+                    getFunctionName(node).c_str(),
+                    TRI_ERROR_QUERY_FUNCTION_ARGUMENT_TYPE_MISMATCH);
+    return AqlValue(AqlValueHintNull());
+  }
+
+  // check that parameters are correct
+  if (scale <= 0 || offset < 0 || decay <= 0 || decay >= 1) {
+    registerWarning(expressionContext,
+                    getFunctionName(node).c_str(),
+                    TRI_ERROR_QUERY_NUMBER_OUT_OF_RANGE);
+    return AqlValue(AqlValueHintNull());
+  }
+
+  // get lambda for further calculation
+  auto decayFunc = decayFuncFactory(origin, scale, offset, decay);
+
+  // argument is number
+  if (argValue.isNumber()) {
+      double arg = argValue.slice().getNumber<double>();
+      double funcRes = decayFunc(arg);
+      return ::numberValue(funcRes, true);
+  } else {
+    // argument is array or range
+    auto* trx = &expressionContext->trx();
+    AqlValueMaterializer materializer(&trx->vpackOptions());
+    VPackSlice slice = materializer.slice(argValue, true);
+    TRI_ASSERT(slice.isArray());
+
+    VPackBuilder builder;
+    {
+      VPackArrayBuilder arrayBuilder(&builder);
+      for (VPackSlice currArg : VPackArrayIterator(slice)) {
+        if (!currArg.isNumber()) {
+          registerWarning(expressionContext,
+                          getFunctionName(node).c_str(),
+                          TRI_ERROR_QUERY_FUNCTION_ARGUMENT_TYPE_MISMATCH);
+          return AqlValue(AqlValueHintNull());
+        }
+        double arg = currArg.getNumber<double>();
+        double funcRes = decayFunc(arg);
+        builder.add(VPackValue(funcRes));
+      }
+    }
+
+    return AqlValue(builder.slice());
```

Comment:
Suggest to use AqlValue ctor from buffer to avoid unnecessary copying 
```suggestion
    return AqlValue(std::move(*builder. steal()));
```